File: shared/spring-base/src/main/java/com/vjcspy/spring/base/dto/response/OkResponse.java
package com.vjcspy.spring.base.dto.response;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

@Getter
@Setter
@Builder(access = AccessLevel.PRIVATE)
public class OkResponse<T> implements BaseResponse<T> {
    @Builder.Default
    private final boolean success = true;

    @Builder.Default
    private final String message = "OK";

    private final T data;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    private final LocalDateTime timestamp;

    private final Map<String, Object> metadata;

    public static <T> OkResponse<T> of(T data) {
        return OkResponse.<T>builder()
                .data(data)
                .timestamp(LocalDateTime.now())
                .metadata(new HashMap<>())
                .build();
    }

    public OkResponse<T> withMetadata(String key, Object value) {
        Map<String, Object> newMetadata = new HashMap<>(this.metadata);
        newMetadata.put(key, value);
        return OkResponse.<T>builder()
                .success(this.success)
                .message(this.message)
                .data(this.data)
                .timestamp(this.timestamp)
                .metadata(newMetadata)
                .build();
    }

    public Map<String, Object> getMetadata() {
        return Collections.unmodifiableMap(metadata);
    }
}

File: shared/spring-base/src/main/kotlin/com/vjcspy/spring/base/dto/response/OkResponse.kt
package com.vjcspy.spring.base.dto.response

import com.fasterxml.jackson.annotation.JsonFormat
import java.time.LocalDateTime
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity

data class OkResponse<T>(
    override val success: Boolean = true,
    override val message: String = "OK",
    override val data: T,
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    val timestamp: LocalDateTime = LocalDateTime.now(),
    val metadata: Map<String, Any> = emptyMap(),
) : BaseResponse<T> {
    companion object {
        fun <T> of(data: T): OkResponse<T> =
            OkResponse(
                data = data,
            )
    }

    fun withMetadata(
        key: String,
        value: Any,
    ): OkResponse<T> =
        copy(
            metadata = metadata + (key to value),
        )
}

/**
 * Extension function to create ResponseEntity from OkResponse
 */
fun <T> OkResponse<T>.toResponseEntity(): ResponseEntity<OkResponse<T>> = ResponseEntity(this, HttpStatus.OK)
